File: rvsec/rvsec-android/rvsmart/src/main/java/br/unb/cic/rvsmart/llm/ToolCallParser.java
package br.unb.cic.rvsmart.llm;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Parses tool calls from LLM responses using a 3-level fallback strategy.
 *
 * Qwen3-VL with SGLang does not always use the native OpenAI tool_calls format (~50% rate).
 * This parser handles all three formats the model may produce:
 *   1. Native  — ChatResponse.toolCalls list is populated by SglangClient
 *   2. XML     — model wraps the call in <tool_call>...</tool_call> tags in its text
 *   3. JSON    — model embeds {"name": "...", "arguments": {...}} in its text
 *
 * Qwen3-VL commonly generates malformed JSON coordinates. A pre-parse fix step (ported
 * from RVAgent's _fix_malformed_json) repairs these before Gson sees the string:
 *   - {"x": 540, 399}      → {"x": 540, "y": 399}   (missing "y" key)
 *   - {"x": [540, 399]}    → {"x": 540, "y": 399}   (array format)
 *   - {"x": .91}           → {"x": 0.91}            (leading-zero float)
 *
 * Action types produced ("click", "long_click", "scroll", "type_text", "back") map
 * directly to Action.Type via AgentLoop conventions.
 */
public class ToolCallParser {

    private static final Gson GSON = new Gson();

    // Matches <tool_call>...</tool_call> or <function_call>...</function_call>
    private static final Pattern XML_TAG_PATTERN = Pattern.compile(
            "<(?:tool_call|function_call)>(.*?)</(?:tool_call|function_call)>",
            Pattern.DOTALL);

    // Malformed JSON fixes (ported from RVAgent tool_call_parser.py _fix_malformed_json)
    // Pattern 1: "x": 352, 782  →  "x": 352, "y": 782  (Qwen3-VL missing "y" key — most common)
    private static final Pattern FIX_MISSING_Y_KEY =
            Pattern.compile("\"x\":\\s*(\\d+),\\s*(\\d+)");
    // Pattern 2: "x": [352, 782]  →  "x": 352, "y": 782  (coordinate array format)
    private static final Pattern FIX_ARRAY_COORDS =
            Pattern.compile("\"x\":\\s*\\[\\s*(\\d+)\\s*,\\s*(\\d+)\\s*\\]");
    // Pattern 3: ": .91  →  ": 0.91  (missing leading zero)
    private static final Pattern FIX_LEADING_ZERO =
            Pattern.compile(":\\s*\\.(\\d+)");

    // Matches a JSON object that has both "name" and "arguments" keys
    private static final Pattern JSON_INLINE_PATTERN = Pattern.compile(
            "\\{[^{}]*\"name\"[^{}]*\"arguments\"[^{}]*(?:\\{[^{}]*\\})[^{}]*\\}|" +
            "\\{[^{}]*\"arguments\"[^{}]*(?:\\{[^{}]*\\})[^{}]*\"name\"[^{}]*\\}",
            Pattern.DOTALL);

    /**
     * Parse a tool call from the model response using native → XML → JSON fallback.
     *
     * @param response the parsed ChatResponse from SglangClient
     * @return a ParsedAction, or null if no recognisable tool call was found
     */
    public ParsedAction parse(SglangClient.ChatResponse response) {
        if (response == null) return null;

        // Level 1: native tool_calls list
        if (response.getToolCalls() != null && !response.getToolCalls().isEmpty()) {
            SglangClient.ToolCall tc = response.getToolCalls().get(0);
            return buildParsedAction(tc.getName(), tc.getArguments());
        }

        // Level 2: XML tag in content text
        String content = response.getContent();
        if (content != null && !content.isEmpty()) {
            ParsedAction fromXml = parseXml(content);
            if (fromXml != null) return fromXml;

            // Level 3: inline JSON in content text
            return parseJsonInline(content);
        }

        return null;
    }

    /**
     * Parse <tool_call>JSON</tool_call> or <function_call>JSON</function_call>.
     */
    private ParsedAction parseXml(String content) {
        Matcher m = XML_TAG_PATTERN.matcher(content);
        if (!m.find()) return null;

        String inner = m.group(1).trim();
        return parseJsonString(inner);
    }

    /**
     * Find and parse the first inline JSON object with "name" + "arguments" keys.
     */
    private ParsedAction parseJsonInline(String content) {
        // Try to find a balanced JSON object containing "name" and "arguments"
        int start = content.indexOf("{");
        while (start >= 0 && start < content.length()) {
            int end = findMatchingBrace(content, start);
            if (end < 0) break;
            String candidate = content.substring(start, end + 1);
            ParsedAction action = parseJsonString(candidate);
            if (action != null) return action;
            start = content.indexOf("{", start + 1);
        }
        return null;
    }

    /**
     * Find the closing brace that matches the opening brace at position 'start'.
     */
    private int findMatchingBrace(String s, int start) {
        int depth = 0;
        for (int i = start; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '{') depth++;
            else if (c == '}') {
                depth--;
                if (depth == 0) return i;
            }
        }
        return -1;
    }

    /**
     * Fix common Qwen3-VL JSON malformations before passing to Gson.
     * Ported from RVAgent tool_call_parser.py _fix_malformed_json().
     *
     * Returns the fixed string, or the original if no fix was needed.
     */
    static String fixMalformedJson(String json) {
        // Fix missing leading zero: ": .91 → ": 0.91
        String s = FIX_LEADING_ZERO.matcher(json).replaceAll(": 0.$1");
        // Fix array coords: "x": [352, 782] → "x": 352, "y": 782
        s = FIX_ARRAY_COORDS.matcher(s).replaceAll("\"x\": $1, \"y\": $2");
        // Fix missing "y" key: "x": 352, 782 → "x": 352, "y": 782
        s = FIX_MISSING_Y_KEY.matcher(s).replaceAll("\"x\": $1, \"y\": $2");
        // Add missing closing braces (truncated JSON)
        int open = 0;
        for (char c : s.toCharArray()) {
            if (c == '{') open++;
            else if (c == '}') open--;
        }
        if (open > 0) {
            StringBuilder sb = new StringBuilder(s);
            for (int i = 0; i < open; i++) sb.append('}');
            s = sb.toString();
        }
        return s;
    }

    /**
     * Parse a JSON string into a ParsedAction if it contains "name" and "arguments".
     * Applies malformed-JSON fixes before parsing (handles Qwen3-VL coordinate malformations).
     */
    private ParsedAction parseJsonString(String json) {
        try {
            // Apply fixes for common Qwen3-VL output malformations before Gson sees the string
            String fixed = fixMalformedJson(json);
            JsonObject obj = GSON.fromJson(fixed, JsonObject.class);
            if (obj == null || !obj.has("name")) return null;

            String name = obj.get("name").getAsString();

            Map<String, Object> args = new java.util.LinkedHashMap<>();
            if (obj.has("arguments")) {
                com.google.gson.JsonElement argsEl = obj.get("arguments");
                if (argsEl.isJsonObject()) {
                    for (Map.Entry<String, com.google.gson.JsonElement> e :
                            argsEl.getAsJsonObject().entrySet()) {
                        com.google.gson.JsonElement v = e.getValue();
                        if (v.isJsonPrimitive()) {
                            com.google.gson.JsonPrimitive prim = v.getAsJsonPrimitive();
                            if (prim.isNumber()) args.put(e.getKey(), prim.getAsDouble());
                            else if (prim.isBoolean()) args.put(e.getKey(), prim.getAsBoolean());
                            else args.put(e.getKey(), prim.getAsString());
                        } else {
                            args.put(e.getKey(), v.toString());
                        }
                    }
                } else if (argsEl.isJsonPrimitive()) {
                    // Arguments embedded as a JSON string
                    return parseJsonString("{\"name\":\"" + name + "\",\"arguments\":" +
                            GSON.fromJson(argsEl.getAsString(), JsonObject.class) + "}");
                }
            }
            return buildParsedAction(name, args);

        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Build a ParsedAction from a parsed action name and its arguments map.
     *
     * The action name maps to the canonical set used by Action.Type:
     *   click, long_click, scroll, type_text, back, etc.
     * Coordinates from Qwen3-VL are in [0, 1000) normalized space;
     * the caller must convert to pixels via CoordinateNormalizer.
     */
    private ParsedAction buildParsedAction(String name, Map<String, Object> args) {
        if (name == null) return null;

        String actionType = name.toLowerCase().replace("-", "_");

        int x = getIntArg(args, "x", 0);
        int y = getIntArg(args, "y", 0);
        String text = getStringArg(args, "text");
        String direction = getStringArg(args, "direction");

        return new ParsedAction(actionType, x, y, text, direction);
    }

    private int getIntArg(Map<String, Object> args, String key, int defaultValue) {
        if (args == null || !args.containsKey(key)) return defaultValue;
        Object val = args.get(key);
        if (val instanceof Number) return ((Number) val).intValue();
        try { return Integer.parseInt(String.valueOf(val)); } catch (Exception e) { return defaultValue; }
    }

    private String getStringArg(Map<String, Object> args, String key) {
        if (args == null || !args.containsKey(key)) return null;
        Object val = args.get(key);
        return val != null ? String.valueOf(val) : null;
    }

    // --- Inner class ---

    /**
     * The result of parsing a tool call from the LLM response.
     *
     * Coordinates (x, y) are in Qwen3-VL normalized [0, 1000) space.
     * Use CoordinateNormalizer to convert to device pixels before execution.
     */
    public static class ParsedAction {
        private final String actionType;
        private final int x;
        private final int y;
        private final String text;
        private final String direction;

        public ParsedAction(String actionType, int x, int y, String text, String direction) {
            this.actionType = actionType;
            this.x = x;
            this.y = y;
            this.text = text;
            this.direction = direction;
        }

        public String getActionType() { return actionType; }
        public int getX() { return x; }
        public int getY() { return y; }
        public String getText() { return text; }
        public String getDirection() { return direction; }

        @Override
        public String toString() {
            return "ParsedAction{type=" + actionType + ", x=" + x + ", y=" + y +
                    ", text=" + text + ", direction=" + direction + "}";
        }
    }
}

File: rv-android/modules/aperv-llm-validation/src/aperv_llm_validation/pipeline/tool_call_parser.py
"""3-level fallback tool call parser — Python replica of ToolCallParser.java (commit b2852dd).

Parses tool calls from LLM responses using native -> XML tag -> inline JSON fallback.
Handles Qwen3-VL malformed JSON coordinates via fix_malformed_json().
"""

from __future__ import annotations

import json
import re

from aperv_llm_validation.data.models import ParsedAction

# Matches <tool_call>...</tool_call> or <function_call>...</function_call>
_XML_TAG_PATTERN = re.compile(
    r"<(?:tool_call|function_call)>(.*?)</(?:tool_call|function_call)>",
    re.DOTALL,
)

# Malformed JSON fixes (ported from Java ToolCallParser / RVAgent tool_call_parser.py)
# Pattern 1: "x": 352, 782  ->  "x": 352, "y": 782
_FIX_MISSING_Y_KEY = re.compile(r'"x":\s*(\d+),\s*(\d+)')
# Pattern 2: "x": [352, 782]  ->  "x": 352, "y": 782
_FIX_ARRAY_COORDS = re.compile(r'"x":\s*\[\s*(\d+)\s*,\s*(\d+)\s*\]')
# Pattern 3: ": .91  ->  ": 0.91
_FIX_LEADING_ZERO = re.compile(r":\s*\.(\d+)")


def fix_malformed_json(s: str) -> str:
    """Fix common Qwen3-VL JSON malformations before parsing.

    Handles:
    - Missing "y" key: ``"x": 352, 782`` -> ``"x": 352, "y": 782``
    - Array coordinates: ``"x": [352, 782]`` -> ``"x": 352, "y": 782``
    - Missing leading zero: ``: .91`` -> ``: 0.91``
    - Truncated JSON: auto-close unbalanced braces
    """
    # Fix missing leading zero
    result = _FIX_LEADING_ZERO.sub(r": 0.\1", s)
    # Fix array coords
    result = _FIX_ARRAY_COORDS.sub(r'"x": \1, "y": \2', result)
    # Fix missing "y" key
    result = _FIX_MISSING_Y_KEY.sub(r'"x": \1, "y": \2', result)
    # Auto-close truncated JSON
    open_count = 0
    for c in result:
        if c == "{":
            open_count += 1
        elif c == "}":
            open_count -= 1
    if open_count > 0:
        result += "}" * open_count
    return result


def parse(response: dict) -> ParsedAction | None:
    """Parse a tool call from an OpenAI-format response dict.

    Uses 3-level fallback:
    1. Native tool_calls array in response["choices"][0]["message"]["tool_calls"]
    2. XML tags (<tool_call>/<function_call>) in content text
    3. Inline JSON with "name" + "arguments" keys in content text

    Returns None if no recognizable tool call was found.
    """
    if response is None:
        return None

    try:
        message = response["choices"][0]["message"]
    except (KeyError, IndexError, TypeError):
        return None

    # Level 1: native tool_calls
    tool_calls = message.get("tool_calls")
    if tool_calls and len(tool_calls) > 0:
        tc = tool_calls[0]
        try:
            name = tc["function"]["name"]
            args_raw = tc["function"].get("arguments", "{}")
            if isinstance(args_raw, str):
                args = json.loads(args_raw)
            else:
                args = args_raw if isinstance(args_raw, dict) else {}
            reasoning = _extract_reasoning(message)
            return _build_parsed_action(name, args, "native", reasoning)
        except (KeyError, json.JSONDecodeError, TypeError):
            pass

    # Level 2 & 3: parse from content text
    content = message.get("content", "")
    if not content:
        return None

    reasoning = _extract_reasoning(message)

    # Level 2: XML tag
    action = _parse_xml(content, reasoning)
    if action is not None:
        return action

    # Level 3: inline JSON
    return _parse_json_inline(content, reasoning)


def _extract_reasoning(message: dict) -> str | None:
    """Extract reasoning field from message if present."""
    # Some models include reasoning in a dedicated field
    reasoning = message.get("reasoning")
    if reasoning:
        return reasoning
    # Check reasoning_content (used by some providers)
    reasoning = message.get("reasoning_content")
    if reasoning:
        return reasoning
    return None


def _parse_xml(content: str, reasoning: str | None) -> ParsedAction | None:
    """Parse <tool_call>JSON</tool_call> or <function_call>JSON</function_call>."""
    m = _XML_TAG_PATTERN.search(content)
    if not m:
        return None
    inner = m.group(1).strip()
    return _parse_json_string(inner, "xml_tag", reasoning)


def _parse_json_inline(content: str, reasoning: str | None) -> ParsedAction | None:
    """Find and parse the first inline JSON object with 'name' + 'arguments' keys."""
    start = content.find("{")
    while 0 <= start < len(content):
        end = _find_matching_brace(content, start)
        if end < 0:
            # Truncated JSON — try the rest of the string; fix_malformed_json will
            # auto-close braces
            candidate = content[start:]
            action = _parse_json_string(candidate, "inline_json", reasoning)
            if action is not None:
                return action
            break
        candidate = content[start : end + 1]
        action = _parse_json_string(candidate, "inline_json", reasoning)
        if action is not None:
            return action
        start = content.find("{", start + 1)
    return None


def _find_matching_brace(s: str, start: int) -> int:
    """Find the closing brace matching the opening brace at position start."""
    depth = 0
    for i in range(start, len(s)):
        if s[i] == "{":
            depth += 1
        elif s[i] == "}":
            depth -= 1
            if depth == 0:
                return i
    return -1


def _parse_json_string(
    json_str: str, parse_level: str, reasoning: str | None
) -> ParsedAction | None:
    """Parse a JSON string into a ParsedAction if it contains 'name' and 'arguments'.

    Applies malformed-JSON fixes before parsing.
    """
    try:
        fixed = fix_malformed_json(json_str)
        obj = json.loads(fixed)
        if not isinstance(obj, dict) or "name" not in obj:
            return None

        name = obj["name"]
        args = {}
        if "arguments" in obj:
            args_raw = obj["arguments"]
            if isinstance(args_raw, str):
                args = json.loads(args_raw)
            elif isinstance(args_raw, dict):
                args = args_raw
        return _build_parsed_action(name, args, parse_level, reasoning)
    except (json.JSONDecodeError, TypeError, ValueError):
        return None


def _build_parsed_action(
    name: str, args: dict, parse_level: str, reasoning: str | None
) -> ParsedAction | None:
    """Build a ParsedAction from a parsed action name and its arguments."""
    if name is None:
        return None

    action_type = name.lower().replace("-", "_")
    x = _get_int_arg(args, "x", 0)
    y = _get_int_arg(args, "y", 0)
    text = args.get("text")
    if text is not None:
        text = str(text)

    return ParsedAction(
        action_type=action_type,
        x=x,
        y=y,
        text=text,
        reasoning=reasoning,
        parse_level=parse_level,
    )


def _get_int_arg(args: dict, key: str, default: int = 0) -> int:
    """Extract an integer argument, handling float and string values."""
    if args is None or key not in args:
        return default
    val = args[key]
    if isinstance(val, (int, float)):
        return int(val)
    try:
        return int(str(val))
    except (ValueError, TypeError):
        return default
